### ISSUE TITLE
Does SAM2 is appropriate for Long Video Like 1Minute~ 5Minute?

### ISSUE THREAD

**Issue by aza1200** [facebookresearch/segment-anything-2#196]
As I execute the SAM2 Model
It takes GPU Memory almost 40GB to 80GB when the data like videos (1 Minute ~ 4Minute)

I tried the init_state's option 
by setting  "offload_video_to_cpu=True"

For Common Computer, It takes lots of memory and costs.
Are there any other way to reduce memory overhead?

**Comment by heyoeyo** [comment]
There are two other [init options](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L43-L45) for reducing VRAM usage: `offload_state_to_cpu` and `async_loading_frames` which might help.

Otherwise, I think the only option to reduce the memory use is to clear out old data, assuming you know in advance that you're not going to re-process prior frames (i.e. you're not going to add new prompts later on in the video and re-run to get a better result). I think that should be possible by adding code to delete old frame data, [just after storing new results](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L716).

Here's a very hacky example (missing checks to make sure there is actually object data to delete):
```python
# (Hacky) Delete old state data to clear space after '_run_single_frame_inference'
storage_key, obj_key = "non_cond_frame_outputs", "output_dict_per_obj"
oldest_allowed_idx = frame_idx - 16
all_frame_idxs = output_dict[storage_key].keys()
old_frame_idxs = [idx for idx in all_frame_idxs if idx < oldest_allowed_idx]
for old_idx in old_frame_idxs:
    output_dict[storage_key].pop(old_idx)
    for objid in inference_state[obj_key].keys():
        inference_state[obj_key][objid][storage_key].pop(old_idx, None)
```
You could even save the .pop() outputs to disk if you didn't want to lose the older results.

Edit: Updated last line as recommended [below](https://github.com/facebookresearch/sam2/issues/196#issuecomment-2881109914).

**Comment by GoldenFishes** [comment]
you can try add `torch.cuda.empty_cache()` after self.memory_attention() in SAM2Base._prepare_memory_conditioned_features

**Comment by ronghanghu** [comment]
Another possible approach to reduce memory consumption is to place the loaded video pixels in half fp16 precision than float32, by changing this line https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/sam2/utils/misc.py#L267 to
```
images = torch.zeros(num_frames, 3, image_size, image_size, dtype=torch.float16) 
```
(i.e. storing the pixels in fp16 format instead of fp32)

**Comment by GoldenFishes** [comment]
> There are two other [init options](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L43-L45) for reducing VRAM usage: `offload_state_to_cpu` and `async_loading_frames` which might help.
> 
> Otherwise, I think the only option to reduce the memory use is to clear out old data, assuming you know in advance that you're not going to re-process prior frames (i.e. you're not going to add new prompts later on in the video and re-run to get a better result). I think that should be possible by adding code to delete old frame data, [just after storing new results](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L716).
> 
> Here's a very hacky example (missing checks to make sure there is actually object data to delete):
> 
> ```python
> # (Hacky) Delete old state data to clear space after '_run_single_frame_inference'
> storage_key, obj_key = "non_cond_frame_outputs", "output_dict_per_obj"
> oldest_allowed_idx = frame_idx - 16
> all_frame_idxs = output_dict[storage_key].keys()
> old_frame_idxs = [idx for idx in all_frame_idxs if idx < oldest_allowed_idx]
> for old_idx in old_frame_idxs:
>     output_dict[storage_key].pop(old_idx)
>     for objid in inference_state[obj_key].keys():
>         inference_state[obj_key][objid][storage_key].pop(old_idx)
> ```
> 
> You could even save the .pop() outputs to disk if you didn't want to lose the older results.

Thank you very much! We used the implementation code you provided and further optimized it by consuming the generated segmentation masks after each processing step. As a result, we were able to achieve a constant memory and GPU usage in our pipeline. 

Here is the link to our repository: https://github.com/motern88/Det-SAM2/tree/main

The related technical report can be found here: https://arxiv.org/abs/2411.18977

**Comment by luisscbb** [comment]
> There are two other [init options](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L43-L45) for reducing VRAM usage: `offload_state_to_cpu` and `async_loading_frames` which might help.
> 
> Otherwise, I think the only option to reduce the memory use is to clear out old data, assuming you know in advance that you're not going to re-process prior frames (i.e. you're not going to add new prompts later on in the video and re-run to get a better result). I think that should be possible by adding code to delete old frame data, [just after storing new results](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L716).
> 
> Here's a very hacky example (missing checks to make sure there is actually object data to delete):
> 
> # (Hacky) Delete old state data to clear space after '_run_single_frame_inference'
> storage_key, obj_key = "non_cond_frame_outputs", "output_dict_per_obj"
> oldest_allowed_idx = frame_idx - 16
> all_frame_idxs = output_dict[storage_key].keys()
> old_frame_idxs = [idx for idx in all_frame_idxs if idx < oldest_allowed_idx]
> for old_idx in old_frame_idxs:
>     output_dict[storage_key].pop(old_idx)
>     for objid in inference_state[obj_key].keys():
>         inference_state[obj_key][objid][storage_key].pop(old_idx)
> You could even save the .pop() outputs to disk if you didn't want to lose the older results.

Bro you're the GOAT. Thanks to this piece of code and comment out the line you mention in issue #264, I was able to track objects in long videos with constant VRAM memory usage. Thank you so much!!!

P.S. you have to modify pop(old_idx) to pop(old_idx, None) in the last line so no KeyError is thrown.

**Comment by dwij-patel123** [comment]
> There are two other [init options](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L43-L45) for reducing VRAM usage: `offload_state_to_cpu` and `async_loading_frames` which might help.
> 
> Otherwise, I think the only option to reduce the memory use is to clear out old data, assuming you know in advance that you're not going to re-process prior frames (i.e. you're not going to add new prompts later on in the video and re-run to get a better result). I think that should be possible by adding code to delete old frame data, [just after storing new results](https://github.com/facebookresearch/segment-anything-2/blob/0db838b11726893f151fa5826ecfa744e1a7760f/sam2/sam2_video_predictor.py#L716).
> 
> Here's a very hacky example (missing checks to make sure there is actually object data to delete):
> 
> # (Hacky) Delete old state data to clear space after '_run_single_frame_inference'
> storage_key, obj_key = "non_cond_frame_outputs", "output_dict_per_obj"
> oldest_allowed_idx = frame_idx - 16
> all_frame_idxs = output_dict[storage_key].keys()
> old_frame_idxs = [idx for idx in all_frame_idxs if idx < oldest_allowed_idx]
> for old_idx in old_frame_idxs:
>     output_dict[storage_key].pop(old_idx)
>     for objid in inference_state[obj_key].keys():
>         inference_state[obj_key][objid][storage_key].pop(old_idx, None)
> You could even save the .pop() outputs to disk if you didn't want to lose the older results.
> 
> Edit: Updated last line as recommended [below](https://github.com/facebookresearch/sam2/issues/196#issuecomment-2881109914).

This solution was awesome for GPU optimization but I don't know why my RAM is getting out of memory when I am using 25GB of RAM on 3min videos even though I am using offload_video_to_cpu=False. Can you help me with the possible issues I might me facing and how can I fix that.

**Comment by heyoeyo** [comment]
> 25GB of RAM on 3min videos

Most likely that would be due to pre-computing/caching all of the frame encodings. There's a way to stop that from happening, but it requires (small) modifications to the code. There's some info about it in issue #264

**Comment by shivanikhandelwal2001** [comment]
I was facing the same GPU memory issue while working with a large dataset, so I went through the code and tried a few things. Sharing what worked and what didn't.

First, I tried processing in batches (chunks of frames) and using the last frame masks as the input for the first frame of the next batch. It works to some extent, but there were a few practical issues:

* I still had to wait for the model to process the entire batch before I could interact with the results.
* If I corrected the mask on any frame, I had to rerun propagation for the remaining frames in that batch.
* VRAM usage was still quite high because the images for the batch were still being loaded and held in memory.

While going through the implementation, one question came to my mind:

> If I am processing frames one by one, why do we need to load all the frames at once?

From what I understood, keeping the frames loaded and keeping the temporal context are two different things.

The model propagates masks using the context stored in structures like:

* `output_dict_per_obj` -  `cond_frame_outputs`, `non_cond_frame_outputs`
* `inputs_per_obj` (point/mask inputs)

These are the memories that are actually required for propagation. The image tensors themselves don't necessarily need to stay loaded for the entire video.

So I modified the video loader instead.

Instead of loading every frame during `init_state()`, I only store the image paths. Whenever the model needs a frame to compute image features, I load that frame, preprocess it, extract the features, and continue. The predictor is initialised only once with the complete video directory, but the frames are loaded lazily based on the current frame index.

```python
  def init_video(self, start_index):
      self.current_frame_idx = start_index
      self.inference_state = self.predictor.init_state(video_path=self.video_dir, offload_video_to_cpu=True)

  @torch.inference_mode()
  def init_state(
      self,
      video_path,
      offload_video_to_cpu=False,
      offload_state_to_cpu=False,
      async_loading_frames=False,
  ):
      """Initialize an inference state."""
      img_mean=(0.485, 0.456, 0.406),
      img_std=(0.229, 0.224, 0.225),
      self.img_mean = torch.tensor(img_mean, dtype=torch.float32).view(1, 3, 1, 1).to(self.device)
      self.img_std = torch.tensor(img_std, dtype=torch.float32).view(1, 3, 1, 1).to(self.device)

      compute_device = self.device  # device of the model
      # images, video_height, video_width = load_video_frames(
      #     video_path=video_path,
      #     image_size=self.image_size,
      #     offload_video_to_cpu=offload_video_to_cpu,
      #     async_loading_frames=async_loading_frames,
      #     compute_device=compute_device,
      # )

      image_paths = get_image_paths(video_path)
      _, video_height, video_width = _load_img_as_tensor(image_paths[0], self.image_size)

      inference_state = {}
      inference_state["images"] = image_paths
      inference_state["num_frames"] = len(image_paths)

  def _get_image_feature(self, inference_state, frame_idx, batch_size):
      """Compute the image features on a given frame."""
      # Look up in the cache first
      image, backbone_out = inference_state["cached_features"].get(
          frame_idx, (None, None)
      )
      if backbone_out is None:
          # Cache miss -- we will run inference on a single image
          device = inference_state["device"]

          image, video_height, video_width = _load_img_as_tensor(inference_state["images"][frame_idx], self.image_size)
          image = image.to(device).float().unsqueeze(0)
          image = (image - self.img_mean) / self.img_std

```

This reduced the VRAM usage significantly in my case. Apart from that, a few other parameters can also help:

* `offload_video_to_cpu=True`
* `predictor.clear_non_cond_mem_around_input=True`
* Enable mixed precision with:

  ```python
  torch.autocast("cuda", dtype=torch.bfloat16).__enter__()
  ```

Also, it's worth checking the `self.num_maskmem` parameter used in `_clear_non_cond_mem_around_input()` - "Remove the non-conditioning memory around the input frame. When users provide correction clicks, the surrounding frames' non-conditioning memories can still contain outdated object appearance information and could confuse the model. This method clears those non-conditioning memories surrounding the interacted frame to avoid giving the model both old and new information about the object."

And yes, clearing old object memories or previous non-conditioning memories (when they are no longer needed) can save a considerable amount of VRAM, depending on the application.

The main change I made was separating frame storage from temporal memory. Instead of keeping all image tensors in the inference state, I only keep the frame paths and load the required frame on demand when image features need to be computed. The temporal memories (cond_frame_outputs, non_cond_frame_outputs, inputs_per_obj) are still maintained for propagation, but the images themselves don't occupy GPU memory throughout the video.

These are just my observations after modifying the loader to handle a larger dataset, happy to know maintainers' thoughts on this.

**Comment by shivanikhandelwal2001** [comment]
**Reduce input resolution:** SAM2 runs the encoder at 1024×1024 by default. Setting image_size: 512 or 768 in the YAML config can give less encoder compute.
